crates/valence_client/src/world_border.rs

```diff
@@ -0,0 +1,373 @@
+pub const DEFAULT_DIAMETER: f64 = (DEFAULT_PORTAL_LIMIT * 2) as f64;
+pub const DEFAULT_WARN_TIME: i32 = 15;
+pub const DEFAULT_WARN_BLOCKS: i32 = 5;
+
+#[derive(SystemSet, Copy, Clone, PartialEq, Eq, Hash, Debug)]
+pub struct UpdateWorldBorderPerInstanceSet;
+
+#[derive(SystemSet, Copy, Clone, PartialEq, Eq, Hash, Debug)]
+pub struct UpdateWorldBorderPerClientSet;
+
+pub(crate) fn build(app: &mut App) {
+    app.configure_set(
+        UpdateWorldBorderPerInstanceSet
+            .in_base_set(CoreSet::PostUpdate)
+            .before(WriteUpdatePacketsToInstancesSet),
+    )
+    .configure_set(
+        UpdateWorldBorderPerClientSet
+            .in_base_set(CoreSet::PostUpdate)
+            .before(FlushPacketsSet),
+    )
+    .add_event::<SetWorldBorderSizeEvent>()
+    .add_systems(
+        (
+            handle_wb_size_change.before(handle_diameter_change),
+            handle_diameter_change,
+            handle_lerp_transition,
+            handle_center_change,
+            handle_warn_time_change,
+            handle_warn_blocks_change,
+            handle_portal_teleport_bounary_change,
+        )
+            .in_set(UpdateWorldBorderPerInstanceSet),
+    )
+    .add_system(handle_border_for_player.in_set(UpdateWorldBorderPerClientSet));
+}
+
+/// A bundle contains necessary component to enable world border.
+/// This struct implements [`Default`] trait that returns a bundle using
+/// Minecraft Vanilla defaults.
+#[derive(Bundle)]
+pub struct WorldBorderBundle {
+    pub center: WorldBorderCenter,
+    pub diameter: WorldBorderDiameter,
+    pub portal_teleport_boundary: WorldBorderPortalTpBoundary,
+    pub warning_time: WorldBorderWarnTime,
+    pub warning_blocks: WorldBorderWarnBlocks,
+    pub moving: MovingWorldBorder,
+}
+
+impl WorldBorderBundle {
+    /// Create a new world border with specified center and diameter
+    pub fn new(center: impl Into<DVec2>, diameter: f64) -> Self {
+        Self {
+            center: WorldBorderCenter(center.into()),
+            diameter: WorldBorderDiameter(diameter),
+            portal_teleport_boundary: WorldBorderPortalTpBoundary(DEFAULT_PORTAL_LIMIT),
+            warning_time: WorldBorderWarnTime(DEFAULT_WARN_TIME),
+            warning_blocks: WorldBorderWarnBlocks(DEFAULT_WARN_BLOCKS),
+            moving: MovingWorldBorder {
+                old_diameter: diameter,
+                new_diameter: diameter,
+                duration: 0,
+                timestamp: Instant::now(),
+            },
+        }
+    }
+}
+
+impl Default for WorldBorderBundle {
+    fn default() -> Self {
+        Self::new([0.0, 0.0], DEFAULT_DIAMETER)
+    }
+}
+
+#[derive(Component)]
+pub struct WorldBorderCenter(pub DVec2);
+
+#[derive(Component)]
+pub struct WorldBorderWarnTime(pub i32);
+
+#[derive(Component)]
+pub struct WorldBorderWarnBlocks(pub i32);
+
+#[derive(Component)]
+pub struct WorldBorderPortalTpBoundary(pub i32);
+
+/// World border diameter can be read by querying
```

Comment:
```suggestion
/// The world border diameter can be read by calling
```
To avoid confusion with ECS queries.